### PR TITLE
azure: support selecting marketplace image in install config

### PR DESCRIPTION
**Pull request by staebler**
Additional fields added to Azure machine pools that give the user the option to select the marketplace image to use for the OS image. The OS image can only be selected for compute machine pools. Validation will be performed that (1) the marketplace image selected exists and (2) that the license terms for the marketplcae image selected have been accepted.

```
osImage:
  publisher: <string>
  offer: <string>
  sku: <string>
  version: <string>
```
    
https://issues.redhat.com/browse/CORS-1824

### CODE REVIEW

**Comment by MayXuQQ** [comment]
@staebler  when the PR is rebased to branch master, please notify me, I can do pre-merge test. thanks a lot.

**Comment by staebler** [comment]
b4fde9775...aaba77d4
* Rebased on changes from https://github.com/openshift/installer/pull/5721

**Comment by patrickdillon** [comment]
nice work!
/approve
/lgtm

**Comment by MayXuQQ** [comment]
@staebler master nodes will support customize the image with market image ? based on my test, just worker nodes use the market image.

**Comment by MayXuQQ** [comment]
test with pr-merge, Failed. 
$ az vm list -g maxu-5799-q44n6-rg -o tsv --query "[].[name,provisioningState]"
maxu-5799-q44n6-master-0	Succeeded
maxu-5799-q44n6-master-1	Succeeded
maxu-5799-q44n6-master-2	Succeeded
maxu-5799-q44n6-worker-southcentralus1-h8sqv	Failed
maxu-5799-q44n6-worker-southcentralus2-7jlnn	Failed
maxu-5799-q44n6-worker-southcentralus3-brfqd	Failed

[kubeconfig](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/ocp-common/job/Flexy-install/94245/artifact/workdir/install-dir/auth/kubeconfig)

**Comment by staebler** [comment]
> @staebler master nodes will support customize the image with market image ? based on my test, just worker nodes use the market image.

Only the worker nodes can use market images.

File: pkg/asset/machines/azure/machines.go [hunk 1]
```diff
@@ -86,6 +86,20 @@ func provider(platform *azure.Platform, mpool *azure.MachinePool, osImage string
 
 	rg := platform.ClusterResourceGroupName(clusterID)
 
+	var image machineapi.Image
+	if mpool.OSImage.Publisher != "" {
+		image.Publisher = mpool.OSImage.Publisher
+		image.Offer = mpool.OSImage.Offer
+		image.SKU = mpool.OSImage.SKU
+		image.Version = mpool.OSImage.Version
```
Comment (patrickdillon) [inline pkg/asset/machines/azure/machines.go:94]: You also need to set `image.Type = machineapi.AzureImageTypeMarketplaceWithPlan`

**Review by patrickdillon — COMMENTED** [review]
/lgtm cancel

File: pkg/asset/machines/azure/machines.go (hunk 1 above)
Comment (staebler) [reply pkg/asset/machines/azure/machines.go:95]: Resolve with aaba77d4a...d930460a7.

**Comment by staebler** [comment]
645815aa7...44e312304
* Added missing descriptions for new Azure client methods.

**Comment by MayXuQQ** [comment]
test pass with installer_payload_image: registry.build01.ci.openshift.org/ci-ln-dk40i2t/release:latest
/label qe-approved
/bugzilla cc-qa

**Comment by MayXuQQ** [comment]
we will support customize the osImage in installconfig.control.platform.azure.osImage, installconfig.compute.platform.azure.osImage and installconfig.platform.azure.defaultMachinePlatform.osImage, even the market Image just can be used in installconfig.compute.platform.azure.osImage , right ? @staebler

**Comment by staebler** [comment]
> we will support customize the osImage in installconfig.control.platform.azure.osImage, installconfig.compute.platform.azure.osImage and installconfig.platform.azure.defaultMachinePlatform.osImage, even the market Image just can be used in installconfig.compute.platform.azure.osImage , right ? @staebler

No. The validation should fail if the fields of `installconfig.controlPlane.platform.azure.osImage` or `installconfig.platform.azure.defaultMachinePlatform.osImage` are set.

**Comment by MayXuQQ** [comment]
> 

I checked, added installconfig.controlPlane.platform.azure.osImage,  then openshift-install create cluster, will 
`ERROR failed to fetch Metadata: failed to load asset "Install Config": failed to create install config: failed to unmarshal install-config.yaml: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal array into Go struct field InstallConfig.controlPlane of type types.MachinePool `
but `.openshift-install explain installconfig.controlPlane.platform.azure.osImage` no error.

**Comment by staebler** [comment]
> > 
> 
> I checked, added installconfig.controlPlane.platform.azure.osImage, then openshift-install create cluster, will `ERROR failed to fetch Metadata: failed to load asset "Install Config": failed to create install config: failed to unmarshal install-config.yaml: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal array into Go struct field InstallConfig.controlPlane of type types.MachinePool ` but `openshift-install explain installconfig.controlPlane.platform.azure.osImage` no error.
> 
> add installconfig.platform.azure.defaultMachinePlatform.osImage then openshift-install create cluster, will `ERROR failed to fetch Metadata: failed to load asset "Install Config": failed to create install config: invalid "install-config.yaml" file: platform.azure.defaultMachinePlatform.osImage: Invalid value: azure.OSImage{Publisher:"redhat", Offer:"ocp-worker", SKU:"ocp-worker-a", Version:"4.8.2021122100"}: cannot specify the OS image for the master machines ` but `openshift-install explain installconfig.platform.azure.defaultMachinePlatform.osImage` no error
> 
> **expect** explain installconfig.controlPlane.platform.azure.osImage , installconfig.platform.azure.defaultMachinePlatform.osImage `FATAL Error executing openshift-install: failed to load schema for the field platform.azure.defaultMachinePlatform.osImage: invalid field osImage, no such property found `

The field exists. It is just that the validation will fail if you try to populate it. For the control plane, I suspect that your install-config.yaml is malformed.

**Comment by Amoghrd** [comment]
Added marketplace images in installconfig.controlPlane.platform.azure.osImage and got `ERROR failed to fetch Metadata: failed to load asset "Install Config": failed to create install config: invalid "install-config.yaml" file: controlPlane.platform.azure.osImage: Invalid value: azure.OSImage{Publisher:"redhat", Offer:"ocp-worker", SKU:"ocp-worker-a", Version:"4.8.2021122100"}: cannot specify the OS image for the master machines`
Works as expected

**Comment by staebler** [comment]
44e312304...768e9148f
* Rebased for changes from #5309.

**Comment by patrickdillon** [comment]
Bad rebase. Close in favor of #5839

**Pull request closed without merging**